x-pack/plugins/fleet/server/services/agents/update_agent_tags_action_runner.ts

```diff
@@ -129,68 +129,80 @@ export async function updateTagsBatch(
 
   appContextService.getLogger().debug(JSON.stringify(res).slice(0, 1000));
 
-  if (options.retryCount === undefined) {
-    // creating an action doc so that update tags  shows up in activity
-    await createAgentAction(esClient, {
-      id: actionId,
-      agents: agentIds,
-      created_at: new Date().toISOString(),
-      type: 'UPDATE_TAGS',
-      total: options.total ?? res.total,
-    });
-    appContextService
-      .getLogger()
-      .debug(
-        `action doc wrote on ${agentIds.length} agentIds, updated: ${res.updated}, failed: ${res.failures}, version_conflicts: ${res.version_conflicts}`
-      );
-  }
-
   // creating unique ids to use as agentId, as we don't have all agent ids in case of action by kuery
   const getUuidArray = (count: number) => Array.from({ length: count }, () => uuidv4());
 
+  const updatedCount = res.updated ?? 0;
+  const updatedIds = getUuidArray(updatedCount);
+
+  const failures = res.failures ?? [];
+
+  const isLastRetry = options.retryCount === MAX_RETRY_COUNT;
+
+  const versionConflictCount = res.version_conflicts ?? 0;
+  const versionConflictIds = isLastRetry ? getUuidArray(versionConflictCount) : [];
+
+  // creating an action doc so that update tags  shows up in activity
+  await createAgentAction(esClient, {
+    id: actionId,
+    agents: updatedIds
+      .concat(failures.map((failure) => failure.id))
+      .concat(isLastRetry ? versionConflictIds : []),
```

Comment:
Tested with this on 50k agents, and it works very well, the `nbAgentsActionCreated` increases in lockstep with `nbAgentsAck` and doesn't exceed 50k. The agents are all updated too when the action moves to `COMPLETE`.


https://user-images.githubusercontent.com/90178898/219608883-527f641c-7f9f-45bc-8425-5119fc747aa1.mov